### ISSUE TITLE
signature error with cyberduck

### ISSUE THREAD

**Issue by KevinWang15** [Mikubill/gofakes3#5]
1. set up `rclone serve s3` with accessKey & secretKey

```
➜  ~ rclone --version
rclone v1.67.0
- os/version: darwin 14.5 (64 bit)
- os/kernel: 23.5.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.22.4
- go/linking: dynamic
- go/tags: none
```

```
➜  ~ cat ~/.config/rclone/rclone.conf

[local]
type = local
```

```
➜  ~ rclone serve s3 --auth-key aaaa,ssss local:/
2024/07/06 01:39:51 NOTICE: Local file system at /: Starting s3 server on [http://127.0.0.1:8080/]
```

2. Open Cyberduck, and install the`S3 (HTTP)` profile in Cyberduck

<img width="600" alt="image" src="https://github.com/rclone/gofakes3/assets/10173472/2e87ed5c-7cce-4c81-8cec-8cb9540483f5">

3. Configure it (I'm sure I have entered the correct accessKey and secretKey)

<img width="476" alt="image" src="https://github.com/rclone/gofakes3/assets/10173472/7b29e86e-3f57-48fa-94a1-410cc3cb17e2">

4. Connect. List bucket works fine
<img width="508" alt="image" src="https://github.com/rclone/gofakes3/assets/10173472/638a0d7a-0368-4180-93a4-e6e45938e6a4">

5. Listing files in a bucket would fail with signature error

<img width="651" alt="image" src="https://github.com/rclone/gofakes3/assets/10173472/052150e7-a6e6-41bc-b56e-c84941cc4936">

---

P.S. using rclone to connect to this S3 server works fine. 

---

cc @ncw

**Comment by ncw** [comment]
Does cyberduck have a debugging log you could show?

Do you know which version of auth cyberduck is using?

Can you run rclone serve s3 with --dump bodies and -vv and post the log of it failing.

Thanks

**Comment by KevinWang15** [comment]
> Does cyberduck have a debugging log you could show?
> 
> Do you know which version of auth cyberduck is using?
> 
> Can you run rclone serve s3 with --dump bodies and -vv and post the log of it failing.
> 
> Thanks

Hi @ncw 

cyberduck uses V4 auth.
I checked its log, it doesn't have much detail about how the signature was calculated (like what the canonicalRequest was).

But I found the solution! (just several minutes ago - what a coincidence)

This fork https://github.com/alist-org/gofakes3 has the code required to make cyberduck V4 auth work (https://github.com/alist-org/gofakes3/commit/5799d51f4555e2f3dbd3bfb93685a3b867754150).

I'm trying to contact the author of this fix as well (https://github.com/alist-org/gofakes3/issues/1#issuecomment-2211755116) cc @itsHenry35

This problem could get fixed soon! 👍👍👍

**Comment by itsHenry35** [comment]
> > Does cyberduck have a debugging log you could show?
> > Do you know which version of auth cyberduck is using?
> > Can you run rclone serve s3 with --dump bodies and -vv and post the log of it failing.
> > Thanks
> 
> Hi @ncw
> 
> cyberduck uses V4 auth. I checked its log, it doesn't have much detail about how the signature was calculated (like what the canonicalRequest was).
> 
> But I found the solution! (just several minutes ago - what a coincidence)
> 
> This fork https://github.com/alist-org/gofakes3 has the code required to make cyberduck V4 auth work ([alist-org@5799d51](https://github.com/alist-org/gofakes3/commit/5799d51f4555e2f3dbd3bfb93685a3b867754150)).
> 
> I'm trying to contact the author of this fix as well ([alist-org#1 (comment)](https://github.com/alist-org/gofakes3/issues/1#issuecomment-2211755116)) cc @itsHenry35
> 
> This problem could get fixed soon! 👍👍👍

Got it, I will try to submit a PR by tomorrow😊

**Comment by KevinWang15** [comment]
Root cause of this is the wrong ordering of query params.

```diff
- req.URL.RawQuery
+ req.URL.Query().Encode() // this ensures the correct ordering
```

Happy to see it fixed in #6
